src/mouse_events.cpp

```diff
@@ -597,6 +597,48 @@ void mouse_handler::mouse_motion(int x, int y, const bool browse, bool update, m
 	}
 }
 
+// Hook for notifying lua game kernel of mouse button events. We pass button as
+// a serpaate argument than the original SDL event in order to manage touch
+// emulation (i.e., long touch = right click) and such.
+bool mouse_handler::mouse_button_event(const SDL_MouseButtonEvent& event, uint8_t button,
+									   map_location loc, bool click)
+{
+	if (gui().view_locked() || button < SDL_BUTTON_LEFT || button > SDL_BUTTON_X2) {
+		return false;
+	} else if (event.state > SDL_PRESSED || !gui().get_map().on_board(loc)) {
+		return false;
+	}
+
+	static const std::string buttons[6] = {
+		[0]					= "",
+		[SDL_BUTTON_LEFT]	= "left",
+		[SDL_BUTTON_MIDDLE]	= "middle",
+		[SDL_BUTTON_RIGHT]	= "right",
+		[SDL_BUTTON_X1]		= "x1",
+		[SDL_BUTTON_X2]		= "x2",
+	};
+	static const std::string events[3] = {
+		[SDL_RELEASED]		= "up",
+		[SDL_PRESSED]		= "down",
+		[2]					= "click",
+	};
+
+	if(game_lua_kernel* lk = pc_.gamestate().lua_kernel_.get()) {
+		lk->mouse_button_callback(loc, buttons[button], events[event.state]);
```

Comment:
It'd be better if this code wouldn't segfault (and either ignore it or throws an exception) when `button` or `event.state` is for whatever reason out of range.

Reply:
edit: okay i just see you check for it above.

Reply:
I think it'd be still better if this was more straightforward to see, especially now that theose constants are no longer used in the array initializer. maybe replace `std::string buttons[6]` with `std::array<std::string, 6> buttons` and compare the parameter then to `buttons.size()` instead of `SDL_BUTTON_X2` (same for `events`)

Reply:
okie dokie

Reply:
We'll just get rid of the events array. It was only ever there for a slight reduction in the construction of std::basic_string objects, which is super-light weight anyway.